crates/services/txpool_v2/src/storage/graph.rs

```diff
@@ -0,0 +1,555 @@
+            );
+            let utxo_id = UtxoId::new(*tx_id, index);
+            match output {
+                Output::Coin { .. } | Output::Change { .. } | Output::Variable { .. } => {
+                    self.coins_creators.remove(&utxo_id);
+                }
+                Output::ContractCreated { contract_id, .. } => {
+                    self.contracts_creators.remove(contract_id);
+                }
+                _ => {}
+            }
+        }
+    }
+
+    fn get_inner(&self, index: &NodeIndex) -> Result<&StorageData, Error> {
+        self.graph
+            .node_weight(*index)
+            .ok_or(Error::TransactionNotFound(format!(
+                "Transaction with index {:?} not found",
+                index
+            )))
+    }
+
+    fn get_dependents_inner(
+        &self,
+        index: NodeIndex,
+    ) -> Result<impl Iterator<Item = NodeIndex> + '_, Error> {
+        Ok(self
+            .graph
+            .neighbors_directed(index, petgraph::Direction::Outgoing))
+    }
+}
+
+impl Storage for GraphStorage {
+    type StorageIndex = NodeIndex;
+
+    fn store_transaction(
+        &mut self,
+        transaction: PoolTransaction,
+        dependencies: Vec<Self::StorageIndex>,
+        collided_transactions: &[Self::StorageIndex],
+    ) -> Result<(Self::StorageIndex, RemovedTransactions), Error> {
+        let tx_id = transaction.id();
+
+        // Add the new transaction to the graph and update the others in consequence
+        let tip = transaction.tip();
+        let gas = transaction.max_gas();
+        let outputs = transaction.outputs().clone();
+
+        // Check if the dependency chain is too big
+        let mut all_dependencies_recursively = HashSet::new();
+        let mut to_check = dependencies.clone();
+        while let Some(node_id) = to_check.pop() {
+            if collided_transactions.contains(&node_id) {
+                return Err(Error::Collided(
+                    "Use a collided transaction as a dependency".to_string(),
+                ));
+            }
+            // Already checked node
+            if all_dependencies_recursively.contains(&node_id) {
+                continue;
+            }
+            let Some(dependency_node) = self.graph.node_weight(node_id) else {
+                return Err(Error::Storage(format!(
+                    "Node with id {:?} not found",
+                    node_id
+                )));
+            };
+            if dependency_node.number_txs_in_chain >= self.config.max_txs_chain_count {
+                return Err(Error::NotInsertedChainDependencyTooBig);
+            }
+            all_dependencies_recursively.insert(node_id);
+            to_check.extend(self.get_dependencies(node_id)?);
+        }
+
+        if all_dependencies_recursively.len() >= self.config.max_txs_chain_count {
+            return Err(Error::NotInsertedChainDependencyTooBig);
+        }
```

Comment:
We can move this check inside of the loop I think

Reply:
Done